File: server/src/scripts/ingest.js
/**
 * @description 文档入库脚本，执行一次即可，知识库更新时重新执行
 * 运行：node src/scripts/ingest.js
 */

import fs, { readFileSync } from "fs";
import path, { join, dirname } from "path";
import { fileURLToPath } from "url";
// import { RecursiveCharacterTextSplitter } from "langchain/text_splitter";
import { RecursiveCharacterTextSplitter } from "@langchain/textsplitters";
import { PGVectorStore } from "@langchain/community/vectorstores/pgvector";
import { Document } from "@langchain/core/documents";
import { embeddings } from "../models/embedding.js";
import { pool } from "../db/postgres.js";

const __dirname = dirname(fileURLToPath(import.meta.url));

const PG_CONFIG = {
  pool,
  tableName: "knowledge_embeddings",
  columns: {
    idColumnName: "id",
    vectorColumnName: "embedding",
    contentColumnName: "content",
    metadataColumnName: "metadata",
  },
};

const readDirOut = (dirPath) => {
  return new Promise((resolve, reject) => {
    fs.readdir(dirPath, (err, files) => {
      if (err) reject(err);
      else resolve(files);
    });
  });
};

// 1. 加载文档
const loadDocs = async () => {
  // 读取指定目录下的文档文件
  // const files = ["products.md", "policies.md"];
  // 动态解析目录下文档
  const dirPath = path.resolve(__dirname, "../data/knowledge");
  const files = await readDirOut(dirPath);

  return files.map((file) => {
    const content = readFileSync(
      join(__dirname, "../data/knowledge", file),
      "utf-8",
    );
    return new Document({
      pageContent: content,
      metadata: { source: file },
    });
  });
};

// 2. 切分文档
// chunkSize：每个切片的最大字符数
// chunkOverlap：相邻切片的重叠字符数，保证语义不在切割处断裂
const splitter = new RecursiveCharacterTextSplitter({
  chunkSize: 500,
  chunkOverlap: 50,
});

// 3. 向量化并存入 pgvector
const ingest = async () => {
  console.log("开始处理文档...");

  const docs = await loadDocs();
  const chunks = await splitter.splitDocuments(docs);
  console.log(`文档切分完成，共 ${chunks.length} 个片段`);

  // 清空旧数据（全量更新场景）
  const client = await pool.connect();

  try {
    await client.query(
      `CREATE TABLE IF NOT EXISTS knowledge_embeddings (
        id       bigserial PRIMARY KEY,
        content  text,
        metadata jsonb,
        embedding vector(1536)
      );`,
    );

    await client.query("TRUNCATE knowledge_embeddings;");
  } finally {
    client.release();
  }

  await PGVectorStore.fromDocuments(chunks, embeddings, PG_CONFIG);

  console.log("文档入库完成");
  await pool.end();
};

ingest().catch((error) => {
  console.error("入库失败：", err.message);
  process.exit(1);
});

File: backend_py/scripts/ingest.py
"""
文档入库脚本，执行一次即可，知识库更新时重新执行
运行：python -m scripts.ingest  （在 server/ 目录下执行）

对照 JS: node src/scripts/ingest.js
"""
import os
from pathlib import Path
from dotenv import load_dotenv
from langchain_text_splitters import RecursiveCharacterTextSplitter
from langchain_community.vectorstores import PGVector
from langchain_core.documents import Document
from models.embedding import embeddings

load_dotenv()

# 数据库连接
PG_USER = os.getenv("PG_USER", "postgres")
PG_PASSWORD = os.getenv("PG_PASSWORD", "123456")
PG_HOST = os.getenv("PG_HOST", "localhost")
PG_PORT = os.getenv("PG_PORT", "5432")
PG_DATABASE = os.getenv("PG_DATABASE", "jisu_ai")
CONNECTION_STRING = f"postgresql://{PG_USER}:{PG_PASSWORD}@{PG_HOST}:{PG_PORT}/{PG_DATABASE}"


def load_docs():
    """1. 加载文档 —— 动态读取 knowledge 目录下所有 .md 文件"""
    knowledge_dir = Path(__file__).parent.parent / "data" / "knowledge"
    docs = []
    for file_path in knowledge_dir.glob("*.md"):
        content = file_path.read_text(encoding="utf-8")
        docs.append(Document(page_content=content, metadata={"source": file_path.name}))
    return docs


def main():
    print("开始处理文档...")

    # 2. 切分文档
    # chunk_size: 每个切片的最大字符数
    # chunk_overlap: 相邻切片的重叠字符数，保证语义不在切割处断裂
    splitter = RecursiveCharacterTextSplitter(
        chunk_size=500,
        chunk_overlap=50,
    )

    docs = load_docs()
    print(f"找到 {len(docs)} 个文档文件: {[d.metadata['source'] for d in docs]}")

    chunks = splitter.split_documents(docs)
    print(f"文档切分完成，共 {len(chunks)} 个片段")

    # 3. 向量化并存入 pgvector
    # PGVector.from_documents 会自动创建表（如果不存在）
    PGVector.from_documents(
        documents=chunks,
        embedding=embeddings,
        connection_string=CONNECTION_STRING,
        collection_name="knowledge",
    )

    print("文档入库完成")


if __name__ == "__main__":
    main()